e2etests/tests/gracefulrestart.go

```diff
@@ -0,0 +1,128 @@
+	"github.com/metallb/frrk8stests/pkg/k8s"
+	"github.com/metallb/frrk8stests/pkg/k8sclient"
+	"github.com/metallb/frrk8stests/pkg/routes"
+	. "github.com/onsi/gomega"
+	frrconfig "go.universe.tf/e2etest/pkg/frr/config"
+	"go.universe.tf/e2etest/pkg/ipfamily"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clientset "k8s.io/client-go/kubernetes"
+)
+
+var _ = ginkgo.Describe("Establish BGP session with EnableGracefulRestart", func() {
+	var (
+		cs       clientset.Interface
+		updater  *config.Updater
+		reporter *k8sreporter.KubernetesReporter
+		nodes    []corev1.Node
+	)
+
+	ginkgo.BeforeEach(func() {
+		var err error
+		updater, err = config.NewUpdater()
+		Expect(err).NotTo(HaveOccurred())
+		reporter = dump.NewK8sReporter(k8s.FRRK8sNamespace)
+
+		for _, c := range infra.FRRContainers {
+			err := c.UpdateBGPConfigFile(frrconfig.Empty)
+			Expect(err).NotTo(HaveOccurred(), "clear any previous config in the infra container failed")
+		}
+		err = updater.Clean()
+		Expect(err).NotTo(HaveOccurred(), "clear any previous config in the API failed")
+
+		cs = k8sclient.New()
+		nodes, err = k8s.Nodes(cs)
+		Expect(err).NotTo(HaveOccurred())
+	})
+
+	ginkgo.AfterEach(func() {
+		if ginkgo.CurrentSpecReport().Failed() {
+			testName := ginkgo.CurrentSpecReport().LeafNodeText
+			dump.K8sInfo(testName, reporter)
+			dump.BGPInfo(testName, infra.FRRContainers, cs)
+		}
+	})
+
+	ginkgo.Context("When restarting the frrk8s deamon pods", func() {
+
+		assertFuncPerEntry := func(ipFam ipfamily.Family, prefix string) {
+			frrs := config.ContainersForVRF(infra.FRRContainers, "")
+			for _, c := range frrs {
+				err := container.PairWithNodes(cs, c, ipFam)
+				Expect(err).NotTo(HaveOccurred(), "set frr config in infra containers failed")
+			}
+
+			peersConfig := config.PeersForContainersWithOptions(frrs, ipFam,
+				config.AllowAllEnabled, config.GracefulRestartEnabled)
+
+			frrConfigCR := frrk8sv1beta1.FRRConfiguration{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "graceful-restart-test",
+					Namespace: k8s.FRRK8sNamespace,
+				},
+				Spec: frrk8sv1beta1.FRRConfigurationSpec{
+					BGP: frrk8sv1beta1.BGPConfig{
+						Routers: []frrk8sv1beta1.Router{
+							{
+								ASN:       infra.FRRK8sASN,
+								Neighbors: peersConfig.FRRK8sNeibghbor(),
+								Prefixes:  []string{prefix},
+							},
+						},
+					},
+				},
+			}
+
+			err := updater.Update(peersConfig.Secrets, frrConfigCR)
+			Expect(err).NotTo(HaveOccurred(), "apply the CR in k8s api failed")
+
+			check := func() error {
+				for _, p := range peersConfig.Peers() {
+					found, err := routes.CheckNeighborHasPrefix(p.FRR, prefix, nodes)
+					Expect(err).NotTo(HaveOccurred())
+					if !found {
+						return fmt.Errorf("Neigh %s does not have prefix %s", p.FRR.Name, prefix)
+					}
+				}
+				return nil
+			}
+
+			Eventually(check, time.Minute, time.Second).ShouldNot(HaveOccurred(),
+				"route should exist before we restart speaker")
```

Comment:
nit: frr-k8s instance instead of speaker

Reply:
done